### PR TITLE
Changelog history service implementation

### PR DESCRIPTION
**Pull request by fbiville**

### CODE REVIEW

**Comment by fbiville** [comment]
@nvoxland I have a preliminary question about the general changelog service work.

Starting with v4+, Neo4j supports multi-tenancy. In Neo4j terms, a server can contain multiple databases. Each database contains its own independent dataset (including indices and constraints).

I was thinking about allowing users to persist the changelog history in 1 database and running the changesets targetting another one. Does this make sense? If so, shouldn't the lock service lock both the "history" database and the target database or?

(Note: this configuration would have to be disabled for Neo4j 3.5, which the plugin needs to support as well).

**Comment by nvoxland** [comment]
That sounds fairly similar to how traditional databases have catalogs/schemas/instances/databases.

What we normally do is just default to "store the databasechangelog and lock tables in your default catalog/schema/whatever". People can have changeSets that make changese to other schemas or catalogs from that same changelog file that uses that single lock table and history table.

The way users end up doing that is either adding catalogName or schemaName attributes to their changeSets to target them to different areas, or using the liquibaseCatalogName or liquibaseSchemaName attribute to change which catalog and/or schema we create the lock and history tables in.

For consistency reasons, it may be best to use the "catalog" terminology for neo4j to mean "database"? We do something similar with other databases that use the "database" term vs. catalog like ms sqlserver.

**Comment by fbiville** [comment]
One thing I'm starting to notice is that a lot of information are available in `RanChangeSet` (which seems to be the Java equivalent of a persisted change set).

In Liqui**graph**, execution contexts and execution types (`MARK_RAN` ... ) are not persisted, they're just XML attributes used later. Is there a specific reason to persist them? Are these included in the checksum computation? If not, what happens if they change?

I see there are now Labels, which seem similar to execution contexts, as you explain [here](https://www.liquibase.org/blog/contexts-vs-labels). Should I support both concepts or only one of them? If both must be supported, which one has precedence in case of conflicting information?

Finally, I'm not completely sure what the following `RanChangeSet` fields are for:

- `changeLog`
- `storedChangeLog`
- `description` (there does not seem to be any equivalent in `ChangeSet`)

I would appreciate it if you can point me to the right direction, @nvoxland 😀
Happy holidays by the way!

**Comment by fbiville** [comment]
How do you deal with SQL injection attacks with changeset tag values?

**Comment by fbiville** [comment]
Re: tagging, I have other questions @nvoxland, given the implementation may slightly vary with Neo4j.
My current implementation creates a `__LiquibaseTag` node, whose only property is the tag value.
It is linked to the last executed `__LiquibaseChangeSet` node.

1. currently, in an empty database, Liquibase creates an empty changeset and tags it with the given tag. 
    * Should I do the same (i.e. create an "empty" `__LiquibaseChangeSet` node to link to)?
    * Should I just create a disconnected `__LiquibaseTag` node? (<-- I'm currently doing this)
    * Should I just ignore the tag?
1. should I allow duplicate tag values in the graph? (<-- I currently don't)
    * if not, what happens if a tag was formerly linked to an "older" `__LiquibaseChangeSet` node and needs to be linked to a more recent `__LiquibaseChangeSet` node? Should I keep the old connection or remove it, i.e. keep the most recent only?

**Comment by nvoxland** [comment]
> One thing I'm starting to notice is that a lot of information are available in `RanChangeSet` (which seems to be the Java equivalent of a persisted change set).
> 
> In Liqui**graph**, execution contexts and execution types (`MARK_RAN` ... ) are not persisted, they're just XML attributes used later. Is there a specific reason to persist them? Are these included in the checksum computation? If not, what happens if they change?
> 
> I see there are now Labels, which seem similar to execution contexts, as you explain [here](https://www.liquibase.org/blog/contexts-vs-labels). Should I support both concepts or only one of them? If both must be supported, which one has precedence in case of conflicting information?
> 
> Finally, I'm not completely sure what the following `RanChangeSet` fields are for:
> 
> * `changeLog`
> * `storedChangeLog`
> * `description` (there does not seem to be any equivalent in `ChangeSet`)
> 
> I would appreciate it if you can point me in the right direction, @nvoxland 😀
> Happy holidays by the way!

It's best to think of RanChangeSet as the java version of what is in the DATABASECHANGELOG table. We're tracking/persisting which changesSets have been executed in there, so the job of the HistoryService is to return back populated RanChangeSet objects corresponding to what was saved in there. Those RanChangeSet objects may or may not correspond to changeSets in the changelog FILE, and the overall/generic liquibase engine will make decisions on what to do based on what comes from the changelog file vs. what RanChangeSets come back from the HistoryService.

So the execType flag, for example, is only something related to the history that RanChangeSet is capturing (was it previously executed, was it re-ran, etc.) and isn't part of the xml defintion of the changeSet.

For checksums, we checksum the changeSet from the file, not the RanChangeSet.  The RanCHangeSet has the lastCheckSum which was what the checksum of the ran changeset was only. So I don't think that question applies.

For labels vs. contexts, I don't think you have to do anything. The standard liquibase logic will handle all that. From a HistoryService standpoint, you will just have your `setExecType()` method called, and you just store that information. If labels or contexts or whatever would make the changeSet not be executed, the setExecType method will never be called.

**Comment by nvoxland** [comment]
> Finally, I'm not completely sure what the following `RanChangeSet` fields are for:
> 
> * `changeLog`
> * `storedChangeLog`
> * `description` (there does not seem to be any equivalent in `ChangeSet`)
> 

There is a constructor to RanChangeSet that takes a given changeSet+execType and copies the corresponding values over. That may help? If you've not looked at StandardChangeLogHistoryService, that would help with usage as well.

ChangeLog and storedChangeLog are storing the file the changeSet was in. We identify the changeSets by path+id+author to make it easy to avoid collisions, and the "changeLog" is the "path" portion of that. Normally, it's the file it's actually in, but to support moving changeSets between files a user can specify a `logicalChangeLog` for the changeSet and so we use that string instead. But, there are a couple edge cases where we need to actually know what file the changeSet was in, so we use the storedChangeLog for that. Most of the time they end up the same, but sometimes not. From a HistoryService standpoint, you should be given that value in the objects coming in, and it's just a question of persisting it and returning it.

Description is a vaguely-human readable version of what is in the changeSet. There is a getDescription() method on changeSet that will return what you should persist.

**Comment by nvoxland** [comment]
> Re: tagging, I have several questions @nvoxland, given the implementation may slightly vary with Neo4j.
> My current implementation creates a `__LiquibaseTag` node, whose only property (called "tag") is the tag value.
> It is linked to the last executed `__LiquibaseChangeSet` node.
> 
> 1. How do you deal with SQL injection attacks with changeset tag values? I cannot use parameterized queries at the moment and I need to make sure the tag value is sanitize before I interpolate it in the Cypher query.
> 2. currently, in an empty database, Liquibase creates an empty changeset and tags it with the given tag.
>    
>    * Should I do the same (i.e. create an "empty" `__LiquibaseChangeSet` node to link to)?
>    * Should I just create a disconnected `__LiquibaseTag` node? (<-- I'm currently doing this)
>    * Should I just ignore the tag?
> 3. should I allow duplicate tag values in the graph? (<-- I currently don't)
>    
>    * if not, what happens if a tag was formerly linked to an "older" `__LiquibaseChangeSet` node and needs to be linked to a more recent `__LiquibaseChangeSet` node? Should I keep the old connection or remove it, i.e. keep the most recent only?
> 4. kinda related: am I supposed to delete all the changelog history when `destroy` is called? Does the deployment ID have any influence on what gets deleted? Tags should be deleted as well, right?

1. For SQL injection, I've honestly not worried about it too much. Liquibase is all about having users specify SQL to execute against the database, and if they wanted to inject something nefarious there would be much more obvious ways to do it... Plus, to support things like updateSQL mode which outputs the SQL to run we can't use prepared statements, so that makes it harder to manage as well. I could see that something like `tag` is a bit of an external value being sent in vs. a changelog, but I'd still not worry too much at this point about sanitizing the input. That is probably left as a later liquibase-general change.

2. I'm fine with the disconnected tag approach. It should be up to the HistoryService to decide what works best to implement the functionality. It should support the ability to tag an empty database, but however you do that is fine.

3. The standard (and therefore expected) behavior currently that `tag EXISTING_TAG` moves the tag from the old location to the new. That is not well documented/defined behavior, but for compatibility reasons it would be best to do that vs. giving a "duplicate tag" validation error. If that is hard to do, though, let me know.

4. Destroy is supposed to remove ALL data related to what was executed. Its used in `dropAll` which first takes out all the non-liquibase-related tables, then removes the history table(s) and finally the lock table(s). It should leave the user with a completely empty database.

**Comment by fbiville** [comment]
>     4. Destroy is supposed to remove ALL data related to what was executed. Its used in `dropAll` which first takes out all the non-liquibase-related tables, then removes the history table(s) and finally the lock table(s). It should leave the user with a completely empty database.

Regarding the lock table, is it also deleted by the `ChangeLogHistoryService` implementation, or does `dropAll` rely on the impl of `LockService#destroy` for that? 

(Which makes me realize that my current impl of `LockService` does not destroy all the persisted tags at the moment, just the one it knows about, I'm going to fix it now)

**Comment by fbiville** [comment]
Another batch of questions 😅 

- Liquibase currently maintains 1 change log history table per schema. Can I run two unrelated series of change sets (in other words if I understand this well: two unrelated change logs) against the same schema today or does that mess up the history table? If not, I'd be biased into creating a single "changeLog" node per database ("upserted" upon initialization).
- Liqui**graph** stores the executed query contents, Liquibase's `RanChangeset` does not seem to do that. I'm wondering whether that would be interesting to do or not?

**Comment by fbiville** [comment]
Another question, following up on your answer on catalog/schema names at the change log/change set level: the JDBC driver for Neo4j allows to select a particular "catalog" via the [connection URL parameter, called database](https://github.com/neo4j-contrib/neo4j-jdbc/#list-of-supported-neo4j-configuration-parameters).

Where/how can I hook that parameter when a specific catalog is specified at the change log or change set level?

File: src/main/java/liquibase/ext/neo4j/database/Neo4jDatabase.java [hunk 1]
```diff
@@ -137,22 +142,30 @@ public String getNeo4jVersion() {
...
-        set(NEO4J_VERSION_ATTRIBUTE, neo4jVersion);
+    public boolean isEnterprise() {
+        return get(NEO4J_EDITION_ATTRIBUTE.toLowerCase(Locale.ENGLISH)).equals("enterprise");
     }
 
-    private void setCatalogSupport(String neo4jVersion) {
+    private void initializeServerAttributes() {
+        Map<String, ?> components = readComponents();
+        String neo4jVersion = (String) components.get("VERSION");
+        set(NEO4J_VERSION_ATTRIBUTE, neo4jVersion);
```
Comment (fbiville) [inline src/main/java/liquibase/ext/neo4j/database/Neo4jDatabase.java:152]: please let me know if I'm abusing `liquibase.database.AbstractJdbcDatabase#attributes` in any way, @nvoxland

**Comment by fbiville** [comment]
@nvoxland Just added a more complete, realistic changelog example to the sources. (Will remove that before the PR is merged).

I've tested so far:
- preconditions
- custom comments
- rollbacks
- tags
- XML and SQL files
- include and includeAll

The only bug I caught was that I reset all the created tags of a given RERAN change set, which ended up disconnecting tags imperatively created with the tag command.

In order to fix this, I decided to differentiate "imperative" (via CLI) tags from the "declarative" ones (via change log files).
I only recreate and relink the latter. The "imperative" tags remain untouched, which leads to the question: is there a command to undo an "imperative" tag?

**Comment by fbiville** [comment]
Now that I have carefully re-read `liquibase.sqlgenerator.core.MarkChangeSetRanGenerator#generateSql` (used via  `MarkChangeSetRanStatement` instantiated in `liquibase.changelog.StandardChangeLogHistoryService#setExecType`) and compared with my implementation, here is what I understand: if the change set to persist contains a `TagDatabaseChange`, then the change set will be inserted or updated with that tag value. As a consequence, if:

1. change set `A` was initially tagged via the CLI tag command (tag value: `"tag1"`) (or not tagged at all)
2. `A` is modified to now also include `<tagDatabase tag="tag2" />`
3. `A` is re-run (`StandardChangeLogHistoryService#setExecType` end up being called)

`A`'s corresponding tag column should be valued with `"tag2"`.

My first implementation regarding tags was to always remove all tag connections to the change set node. 
However, that would not work in scenarios such as "`A` was tagged with CLI" and is re-run **without**  `<tagDatabase />`. Indeed, tags created via the CLI are not automatically populated in `liquibase.changelog.ChangeSet#changes` and I would end up untagging `A`!

My second (current) implementation is also wrong: it removes only the tags only created via `<tagDatabase />` and relinks to the current `TagDatabaseChange` of `A`. But if `A` was first tagged with the CLI, that tag should be unlinked and the new tag from `TagDatabaseChange` should be linked instead.

What I should do instead: if **and only if** `A` has a `TagDatabaseChange`, untag `A` and upsert the tag and link it to `A`.

WDYT @nvoxland?

**Comment by fbiville** [comment]
That also makes me realize two discrepancies between my `tag` implementation and the standard one.

### First discrepancy: untag the change set first
If the most recent change set is already tagged, it should be untagged first before the new tag gets linked to it. That is not currently the case in my impl.

### Tag value can only be set to a single change set
This is what my implementation currently makes sure of. However, in the standard implementation (looking at `liquibase.sqlgenerator.core.TagDatabaseGenerator`), I do not see where this is enforced.

In the following scenario with Liquibase, I believe the same tag value could tag 2 different change sets:

1. use the CLI to tag the most recent change set (at the time, let's say it's change set `A`) with the value `"tag"`
2. repeat the same operation, still with `"tag"`, but this time the most recent change set is `B`

In that case, unless I'm missing something, both change sets `A` and `B` are tagged with `"tag"`.
My assumption was that a given tag value could only tag a single change set.

Did I overlook something, @nvoxland?

**Comment by fbiville** [comment]
@nvoxland I have two basic end-to-end tests now: 1 for `updateSQL`, 1 for `update`.
I think the PR is good for review now, I don't plan to push any more changes.

**Comment by nvoxland** [comment]
Changes all look good. Merging it in, thanks!

**Comment by fbiville** [comment]
Thanks a lot @nvoxland!

**Pull request merged**